pkg/util/limiter/utilization.go

```diff
@@ -0,0 +1,330 @@
+	memR, err := readFileNoStat(memPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer memR.Close()
+	// Summing RSS and cache for memory usage
+	// For reference, see section 5.5 in https://www.kernel.org/doc/Documentation/cgroup-v1/memory.txt
+	var memUtil uint64
+	var numMemStats int
+	for {
+		var name string
+		var val uint64
+		if _, err := fmt.Fscanf(memR, "%s %d\n", &name, &val); err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+
+			return 0, 0, errors.Wrapf(err, "failed scanning %s", memPathV1)
+		}
+
+		if name == "rss" || name == "cache" {
+			memUtil += val
+			numMemStats++
+			if numMemStats == 2 {
+				break
+			}
+		}
+	}
+	if numMemStats != 2 {
+		return 0, 0, fmt.Errorf("failed scanning %s", memPathV1)
+	}
+
+	// For reference, see https://www.kernel.org/doc/Documentation/cgroup-v1/cpuacct.txt
+	cpuR, err := readFileNoStat(cpuPathV1)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer cpuR.Close()
+	// CPU time in nanoseconds
+	var cpuTime uint64
+	if _, err := fmt.Fscanf(cpuR, "%d", &cpuTime); err != nil {
+		return 0, 0, errors.Wrapf(err, "failed scanning %s", cpuPathV1)
+	}
+
+	return float64(cpuTime) / 1e09, memUtil, nil
+}
+
+func newCgroupV1Scanner() (utilizationScanner, bool) {
+	if _, err := os.Stat(memPathV1); err != nil {
+		return cgroupV1Scanner{}, false
+	}
+	if _, err := os.Stat(cpuPathV1); err != nil {
+		return cgroupV1Scanner{}, false
+	}
+	return cgroupV1Scanner{}, true
+}
+
+type procfsScanner struct {
+	proc procfs.Proc
+}
+
+func (s procfsScanner) Method() string {
+	return "/proc"
+}
+
+func (s procfsScanner) Scan() (float64, uint64, error) {
+	ps, err := s.proc.Stat()
+	if err != nil {
+		return 0, 0, errors.Wrap(err, "failed to get process stats")
+	}
+
+	return ps.CPUTime(), uint64(ps.ResidentMemory()), nil
+}
+
+func newProcfsScanner() (utilizationScanner, bool) {
+	p, err := procfs.Self()
+	if err != nil {
+		return procfsScanner{}, false
+	}
+
+	return procfsScanner{
+		proc: p,
+	}, true
+}
+
+// UtilizationBasedLimiter is a Service offering limiting based on CPU and memory utilization.
+//
+// The respective CPU and memory utilization limits are configurable.
+type UtilizationBasedLimiter struct {
+	services.Service
+
+	logger             log.Logger
+	utilizationScanner utilizationScanner
+
+	// Memory limit in bytes
+	memoryLimit uint64
+	// CPU limit in cores
+	cpuLimit float64
+	// Last CPU time counter
+	lastCPUTime float64
+	// The time of the last update
+	lastUpdate     time.Time
+	movingAvg      *math.EwmaRate
+	LimitingReason atomic.String
+}
+
+// NewUtilizationBasedLimiter returns a UtilizationBasedLimiter configured with cpuLimit and memoryLimit.
+func NewUtilizationBasedLimiter(cpuLimit float64, memoryLimit uint64, logger log.Logger) *UtilizationBasedLimiter {
+	// Calculate alpha for a minute long window
+	// https://github.com/VividCortex/ewma#choosing-alpha
+	alpha := 2 / (60/resourceUtilizationUpdateInterval.Seconds() + 1)
```

Comment:
Could this be a constant?

Reply:
There is a practical drawback to making it a constant, which is that you have to duplicate `resourceUtilizationUpdateInterval` in the formula. Meaning that if you were to change `resourceUtilizationUpdateInterval`, you would also have to remember to change the `alpha` formula correspondingly. I don't see it being worth it.